pkg/sql/schemachanger/scplan/internal/scgraph/dep_edge_tree.go

```diff
@@ -11,100 +11,217 @@
+// getTargetIdxFunc is used to get the location of the node in the graph.
+type getTargetIdxFunc = func(n *screl.Node) targetIdx
+
+// depEdges is a data structure to store the set of depEdges. It offers
+// in-order traversal of edges from or to any node in the graph.
+type depEdges struct {
+	fromTo *btree.BTree
+	toFrom *btree.BTree
+
+	getTargetIdx getTargetIdxFunc
+	edgeAlloc    depEdgeAlloc
+	entryAlloc   depEdgeTreeEntryAlloc
+}
+
+// makeDepEdges constructs the depEdge structure.
+func makeDepEdges(getTargetIdx getTargetIdxFunc) depEdges {
+	const degree = 32 // arbitrary
+	return depEdges{
+		fromTo:       btree.New(degree),
+		toFrom:       btree.New(degree),
+		getTargetIdx: getTargetIdx,
+	}
 }
 
-func (et *depEdgeTree) insert(e *DepEdge) {
-	et.t.ReplaceOrInsert(&edgeTreeEntry{
-		t:    et,
-		edge: e,
+// insertOrUpdate will insert a new dep edge if no such edge exists between
+// from and to. Otherwise, it will update the edge accordingly. An error
+// may be returned if the kind is incompatible with the existing kind for
+// the edge.
+func (et *depEdges) insertOrUpdate(rule Rule, kind DepEdgeKind, from, to *screl.Node) error {
+	k := makeEdgeKey(et.getTargetIdx, from, to)
+	if got, ok := et.get(k); ok {
+		return updateExistingDepEdge(rule, kind, got)
+	}
+	de := et.edgeAlloc.new(DepEdge{
+		kind:  kind,
+		from:  from,
+		to:    to,
+		rules: []Rule{rule},
 	})
+	et.fromTo.ReplaceOrInsert(et.entryAlloc.new(k, fromTo, de))
+	et.toFrom.ReplaceOrInsert(et.entryAlloc.new(k, toFrom, de))
+	return nil
 }
 
-func (et *depEdgeTree) get(e *DepEdge) *DepEdge {
-	got, ok := et.t.Get(&edgeTreeEntry{
-		t:    et,
-		edge: e,
-	}).(*edgeTreeEntry)
-	if !ok {
-		return nil
+func updateExistingDepEdge(rule Rule, kind DepEdgeKind, got *DepEdge) error {
+	if got.kind != kind && kind != Precedence {
+		if got.kind != Precedence {
+			return errors.AssertionFailedf(
+				"inconsistent dep edge kinds: %s rule %q conflicts with %s",
+				rule.Kind, rule.Name, got,
+			)
+		}
+		got.kind = kind
 	}
-	return got.edge
+	got.rules = append(got.rules, rule)
+	return nil
+}
+
+// iterateFromNode iterates the set of edges from the passed node.
+func (et *depEdges) iterateFromNode(n *screl.Node, it DepEdgeIterator) (err error) {
+	return iterateDepEdges(
+		et.getTargetIdx(n), n.CurrentStatus, fromTo, et.fromTo, it,
+	)
+}
+
+// iterateFromNode iterates the set of edges to the passed node.
+func (et *depEdges) iterateToNode(n *screl.Node, it DepEdgeIterator) (err error) {
+	return iterateDepEdges(
+		et.getTargetIdx(n), n.CurrentStatus, toFrom, et.toFrom, it,
+	)
 }
 
-func (et *depEdgeTree) iterateSourceNode(n *screl.Node, it DepEdgeIterator) (err error) {
-	e := &edgeTreeEntry{t: et, edge: &DepEdge{}}
-	if et.order == fromTo {
-		e.edge.from = n
-	} else {
-		e.edge.to = n
+func iterateDepEdges(
+	target targetIdx, status scpb.Status, order edgeTreeOrder, t *btree.BTree, it DepEdgeIterator,
+) (err error) {
+	var idx int
+	if order == toFrom {
+		idx = 1
```

Comment:
it seems that here, we are using int `0` and `1` to represent whether we are talking about the `from` Btree or the `to` Btree, right?